### DISCUSSION TITLE
Pixmap.copy not reflected in saved document

### DISCUSSION THREAD

**Discussion by python3Berg** [pymupdf/PyMuPDF#1495] [Q&A]
I am trying to use unpaper to preprocess pdf images before sending them to ocr. Following the documentation examples, it appears I should be able to copy a deskewed pixmap into the existing document and then have a cleaner page for subsequent processing. Following the example almost verbatim and then saving document at end of page loop, I would expect to see something similar to deskewing provided by ocrmypdf or equivalent. Instead I see document that appears unchanged. I am far more experienced with extraction than modification so perhaps I am missing something obvious. The preprocessing is working..just not saving

```
path = "/media/sf_Home/training documents/bad docs/test.pdf"
files = glob.glob(path)

for fc, fileName in enumerate(files):
    tempdir = "/media/sf_Home/"
    tmppath = os.path.join(tempdir, str(fc))
    if not os.path.exists(tmppath):
        os.makedirs(tmppath)

    try:
        print("Processing ", fileName)
        doc = fitz.open(fileName)  # open document
    except:
        e = sys.exc_info()[0]

    for pg_no, page in enumerate(doc):
        inname = os.path.join(tmppath, str(pg_no) + "-in.ppm")
        outname = os.path.join(tmppath, str(pg_no) + "-out.ppm")

        pix = page.getPixmap()
        pix.save(inname)
        subprocess.call(
            [
                "unpaper",
                inname,
                outname,
            ]
        )
        newPix = fitz.Pixmap(outname)
        pix.setOrigin(0, 0)
        newPix.setOrigin(0, 0)
        pix.copy(newPix, (0, 0, pix.width, pix.height))
        os.remove(inname)
        os.remove(outname)
    doc.save(doc.name, incremental=True, encryption=fitz.PDF_ENCRYPT_KEEP)

```

## Your configuration (mandatory)
Ubuntu 20.04
python 3.8
pymupdf 1.18.17

**Comment by JorjMcKie** [comment]
taking the liberty to transform this into a **Discussions** item ...

**Comment by JorjMcKie** [comment]
I don't know about "unpaper", but it seems it does somethinbg with an image to produce a new image.
So far so good.
But then I don't see you do anything with the result. You could insert it as a page in a PDF, or convert it to an OCRed page, ...
For example, convert the resulting PPM image to a (1-page) PDF document and append that PDF to an existing other PDF.

```python
imgdoc = fitz.open(newpix.tobytes())  # make a Document from the pixmap's image
pdfbytes = imgdoc.convert_to_pdf()  # convert it to the binary stream of a PDF
imgpdf = fitz.open("pdf", pdfbytes)  # make a PDF doc from the stream data
doc.insert_pdf(imgpdf)  # append the original image as a PDF page
# clean up stuff
imgdoc.close()
imgpdf.close()
newpix = None
```
As I said: just an example. You could have OCRed newpix first too or whatever

**Reply by python3Berg** [reply]
Maybe this is what I am missing...I thought I was copying newPix over pix, in effect overwriting the original page image. The documentation showed a modified page with a clip of a dog overwriting original text. I guess I wrongly assuming we were modifying the document instead of just manipulating images. If that is an incorrect assumption, I will need to revisit whether this is anymore efficient than what's already in place. Thanks...

**Reply by JorjMcKie** [reply]
Ok, I see now where you took the wrong turn.
Whatever "unpaper" does, its result is an image which you must bring back to some PDF document one way or another.
The original PDF, from where the pages were taken to make images from, does not take notice of any of that doing. It wouldn't even "know" that an image was made from its pages - these are totally disconnected things ...

You seem to try making high quality OCRs from PDF pages.
Why don't yu simply use the new OCR feature of PyMuPDF? You control the OCR quality of a page by setting a high dpi value. Or let PyMuPDF detect which parts of a page need to be OCRed at all (you may have mixtures normal text and text in images), etc.
Would save you using unpaper via triggering a subprocess.

**Reply by python3Berg** [reply]
Tesseract has some significant issues wrt resulting coordinates. I have been using it even before use of pymupdf and have found that the google cloud vision and aws textract yield better, faster results. Unpaper is used by ocrmypdf that you reference to deskew and clean pages. This is the missing part for me and I am only calling this once I already know if ocr is needed. Thanks for the help. I do find it confusing that there is a whole object and section doing image processing that has no impact on the pdf.
